ConditionalCompilation.md

```diff
@@ -1,28 +1,140 @@
+
+### Location of *Compile-time attributes*
+
+(TODO: this needs more reflexion)
+
+A Compile-time attribute CAN decorate the following syntax nodes:
+ * structure declarations
+ * structure members
+ * function declarations
+ * ?? function formal parameter declarations ?? (is this needed?)
+ * variable declarations and value declarations
+ * all statements, except those disallowed below
+ * directives
+
+Because eliminating the decorated node would lead to invalid code, a Compile-time attribute CANNOT decorate the following syntax nodes:
+ * bodies of function declarations, switch statements, switch clauses, loop statements, for statements, while statements, if/else statements
+ * function return types
+ 
+> Note: The WGSL grammar currently does not allow attributes in front of const value declarations, variable declarations, directives, switch clauses and several statements. We would extend the syntax to allow them.
+> Which exacly we will enable is subject to discussion. Enabling them before all statements is perhaps undesirable.
+
+### Syntax 1: Features only and the `@ifdef`, `@ifndef` attributes
+
+> This is the least complex implementation.
+
+A *compile-time expression* must be a single *compile-time feature*.
+
+Two new attributes are introduced that take a single parameter.
+* The `@ifdef` attribute marks the decorated node for removal if the parameter evaluates to `false`.
+* The `@ifndef` attribute marks the decorated node for removal if the parameter evaluates to `true`.
+
+*Example*
+
+```rs
+@ifdef(my_feature)
+fn f() { ... }
+@ifndef(my_feature)
+fn f() { ... }
+```
+
+### Syntax 2: `@if` attribute with compile-time logical expressions
+
+> This is the medium-complexity implementation.
+
+A *compile-time expression* is a subset of normal WGSL [expressions](https://www.w3.org/TR/WGSL/#expressions). it must be one of:
+ * a *compile-time feature*,
+ * a [logical expression](https://www.w3.org/TR/WGSL/#logical-expr): logical not (`!`), short-circuiting AND (`&&`), short-circuiting OR (`||`),
+ * a [parenthesized expression](https://www.w3.org/TR/WGSL/#parenthesized-expressions).
+
+Three new attributes are introduced.
+* An `@if` attribute takes a single parameter. It marks the decorated node for removal if the parameter evaluates to `false`.
+* An `@elif` attribute decorates the next sibling of a syntax node decorated by a `@if` or an `@elif`. It takes one parameter.
+   It marks the decorated node for removal if its parameter evaluates to `true` AND if the previous `@if` and `@elif` attribute parameters evaluate to false.
+* An `@else` attribute decorates the next sibling of a syntax node decorated by a `@if` or an `@elif`. It does not take any parameter.
+   It marks the decorated node for removal if the previous `@if` and `@elif` attribute parameters evaluate to false.
+
+*Example*
+
+```rs
+@if(feature_1 && (!feature_2 || feature_3))
+fn f() { ... }
+@elif(!feature_1)
+fn f() { ... }
+@else
+fn f() { ... }
+```
```

Comment:
hmm. nice that it's familiar to users of #if #else #endif. 
but also doesn't add much vs. #if #else #endif...

Reply:
IMO we don't really need `@elif` and `@else` in the simplest variant of the proposal. I believe they also complicate the parsing a fair bit, since the following should be illegal.
```
@if(feature_1)
struct Foo { a: f32}
struct Bar { a: f32 }
@else
struct Foo { a: vec3f }
```

Reply:
@stefnotch that would not be caught at parse time, more likely as a later contextual check. 
but yeah I agree that elif else are not necessary at first. That was also part of the idea of proposal 1 with only ifdef and ifndef.

I'd say we can see how hard it is to implement them. I have the feeling that it's not too much additional work to support them.

Reply:
+1 for starting simple if it seems sufficient to you.